eth/downloader/downloader.go

```diff
@@ -1017,7 +1031,7 @@ func (d *Downloader) findAncestor(p *peer) (uint64, error) {
 					continue
 				}
 				// Otherwise check if we already know the header or not
-				if (d.mode != LightSync && d.hasBlockAndState(headers[i].Hash())) || (d.mode == LightSync && d.hasHeader(headers[i].Hash())) {
+				if (d.mode == FullSync && d.hasBlockAndState(headers[i].Hash())) || (d.mode != FullSync && d.hasHeader(headers[i].Hash())) {
```

Comment:
`!= LightSync` meand `FastSync || FullSync`. However `hasBlockAndState` is only meaningful for `FullSync` since fast sync does not have the state yet. This basically meant that when fast sync resumed a previous sync, it always skipped this fast ancestor detection and had to do the binary search version.